### PR TITLE
Let get_frame_attr_names only return names, add get_frame_attr_defaults

### LINKED ISSUE #13483: `get_frame_attr_names` is misleadingly named and a performance issue
### Description
One more from my profiling...

`BaseFrame.get_frame_attr_names` builds a new dict every time of `name, getattr(cls, name)`. Which is 
a) misleading, since it does not only return the *names* 
b) a performance issue, since code in `SkyCoord` for example does membership tests on `get_frame_attr_names()`. 

### Expected behavior

A function `get_frame_attr_names` should only return frame attribute names, not a mapping of frame attributes to defaults.

### Actual behavior

Membership tests are much more expensive then they need to be like this.

### PR DESCRIPTION
**Pull request by maxnoe**
### Description

I let `get_frame_attr_names` really only return the names, not build a dict.

At the few places where really the dict is used, i added the (basically renamed) `get_frame_attr_defaults`) function.

Fixes #13483 

I had to add a call to `get_frame_attr_defaults()` to `BaseCoordinateFrame.__init__subclass__` as this was previously relied on to initialize the values.

Benchmark script:

```
import astropy.units as u
from astropy.coordinates import EarthLocation, AltAz, SkyCoord
from astropy.time import Time

loc = EarthLocation.of_site("Roque de los Muchachos")
time = Time.now()
frame = AltAz(location=loc, obstime=time)
coord = SkyCoord(0 * u.deg, 0 * u.deg, frame=frame)

for attr in ('obstime', 'location', 'temperature'):
    print(attr)
    get_ipython().run_line_magic('timeit', f'coord.{attr}')
```

Results on main:

```
obstime
8.28 µs ± 15.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
location
8.61 µs ± 342 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
temperature
8.76 µs ± 186 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Results here:

```
obstime
3.24 µs ± 15.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
location
2.32 µs ± 30.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
temperature
3.56 µs ± 44.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```
```


### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: docs/coordinates/frames.rst
Comment (pllim) [inline docs/coordinates/frames.rst:50]:
```suggestion
```

**Review by pllim — COMMENTED** [review]
Need a change log.

As with your other PRs, feel free to ping relevant maintainer(s) to review. Thanks!

**Comment by pllim** [comment]
The RTD failure is transient and would hopefully go away on your next push.

The `astropy/coordinates/tests/test_arrays.py::test_angle_arrays` failure is unrelated.

File: docs/coordinates/frames.rst
Comment (maxnoe) [reply docs/coordinates/frames.rst:50]: Fixed (Note that github didn't pick it up, seems it was confused by the two blank lines going to one but thinking I removed the *wrong* blank line.

**Comment by WilliamJamieson** [comment]
Cause of your failing CI was just fixed. Please rebase.

File: docs/coordinates/frames.rst
Comment (astrofrog) [inline docs/coordinates/frames.rst:49]: Should you actually cast to a list or tuple to avoid having this be a dict_keys object?

**Comment by astrofrog** [comment]
@maxnoe - since this is a change in API, we should get this in today or tomorrow if possible. I think this can be treated as a bug really since indeed what was being returned was not what the method is called, and I'm not sure how we would deprecate the current behavior so I think we should just bite the bullet and make the change in 5.2.

File: docs/coordinates/frames.rst
Comment (maxnoe) [reply docs/coordinates/frames.rst:49]: What is the disadvantage of that? It prints nicely, supports all the operations (most of them even faster then `tuple`) and doesn't make a copy necessary in performance critical code

Comment (eerovaher) [reply docs/coordinates/frames.rst:49]: An argument can be made that the result here should be a `list` because of consistency with `Table.colnames`, which also returns a `list`. `dict_keys` could be useful in many cases, but they can be accessed directly as `frame_attributes.keys()`, similarly to what `Table.columns.keys()` provides.

Comment (maxnoe) [reply docs/coordinates/frames.rst:49]: But then, for the mentioned performance benefits, all code currently using `get_frame_attr_names` should just use `frame_attributes.keys()` and there is actually no real reason to have `get_frame_attr_names` at all.

Comment (maxnoe) [reply docs/coordinates/frames.rst:49]: All places where `get_frame_attr_names` is used internally in astropy are two cases:

a) iterating over the names
b) membership testing

both are faster when not making a list copy of the available dict keys

Comment (eerovaher) [reply docs/coordinates/frames.rst:49]: If we do decide to replace all uses of `get_frame_attr_names()` with `frame_attributes.keys()` then that would at least provide a clean way of deprecating the former.

Comment (maxnoe) [reply docs/coordinates/frames.rst]: done

File: docs/coordinates/skycoord.rst
Comment (eerovaher) [inline docs/coordinates/skycoord.rst:552]: The text here shouldn't refer to key values anymore.

File: docs/coordinates/frames.rst
Comment (eerovaher) [inline docs/coordinates/frames.rst:52]: Is there a reason not to provide a link to the method?
```suggestion
:meth:`~astropy.coordinates.Baseframe.get_frame_attr_defaults`::
```

File: astropy/coordinates/baseframe.py
Comment (eerovaher) [inline astropy/coordinates/baseframe.py:704]: It would be better if these methods had docstrings, but the because there weren't any before either then this shouldn't necessarily prevent merging.

**Review by eerovaher — COMMENTED** [review]
If we decide that `get_attr_names()` should exist as a method that returns a `list` then we are almost done, but the documentation could benefit from a couple of small changes.

However, if we decide that `get_frame_attr_names()` serves no purpose anymore then it should be deprecated properly. I would prefer the latter.

**Comment by maxnoe** [comment]
> However, if we decide that get_frame_attr_names() serves no purpose anymore then it should be deprecated properly. I would prefer the latter.

What would need to be done for this?

**Comment by eerovaher** [comment]
You could use the `@deprecated` decorator from `astropy.utils.decorators`, like in e.g. https://github.com/astropy/astropy/blob/c660b079b6472920662ca4a0c731751a0342448c/astropy/cosmology/realizations.py#L81-L83
Deprecating would have the advantage that it wouldn't break any code, whereas changing the method so that it returns a list of attribute names might.

**Comment by maxnoe** [comment]
Ok, so you mean I should leave the old method untouched, replace internal usage with just using `frame_attributes` and add the `get_frame_attr_defaults` method?

**Comment by eerovaher** [comment]
Yes, all internal usage should be replaced and the old method should be left untouched, it would just have to be decorated and the name of the new method should be provided as `alternative`.

**Comment by maxnoe** [comment]
Sorry, but looking at the places this is used and at the comments, it seems like this is a major api change.... not just the simple performance improvement I had in mind.

Making `get_frame_attribute_names` fast and use it as intended seems much less intrusive than completely not using it anymore....

However, one example of a complex project always brought up with coordinate changes is sunpy, and they don't have custom `get_frame_attribute_names`, so maybe it's ok?

We also certainly are not using it in ctapipe.

**Comment by maxnoe** [comment]
Speedups with respect to master are still about the same:

```
❯ ipython -i time_attrs.py
Python 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:22:55) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.2.0 -- An enhanced Interactive Python. Type '?' for help.
obstime
4.79 µs ± 423 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
location
2.85 µs ± 52.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
temperature
4.35 µs ± 62.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

❯ git switch main
Switched to branch 'main'
Your branch is up to date with 'origin/main'.
❯ ipython -i time_attrs.py
Python 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:22:55) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.2.0 -- An enhanced Interactive Python. Type '?' for help.
obstime
13 µs ± 218 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
location
12 µs ± 461 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
temperature
13.8 µs ± 337 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

File: astropy/coordinates/baseframe.py
Comment (eerovaher) [inline astropy/coordinates/baseframe.py:703]: `alternative` should be the actual alternative of the method, not the alternative of what the deprecated method name would suggest. If you do wish to explicitly mention how to get the actual frame attribute names then use a custom warning message with the `message` argument.
```suggestion
    @deprecated("5.2", alternative="get_frame_attr_defaults")
```

File: docs/changes/coordinates/13484.bugfix.rst
Comment (eerovaher) [inline docs/changes/coordinates/13484.bugfix.rst:3]: Did something go wrong during the rewrite? And there is no need to mention that internal usage has changed, that is an implementation detail the end users should not care about.

File: docs/coordinates/skycoord.rst
Comment (eerovaher) [inline docs/coordinates/skycoord.rst:550]: Would it be better to demonstrate the full `dict` here not just the keys?

**Review by eerovaher — COMMENTED** [review]
Deprecating the old method instead of changing its output is a much less disruptive change. I have a couple remarks about documentation, and I would recommend squashing the commits.

File: docs/changes/coordinates/13484.bugfix.rst
Comment (maxnoe) [reply docs/changes/coordinates/13484.bugfix.rst:3]: yes, fixed

File: docs/coordinates/skycoord.rst
Comment (maxnoe) [reply docs/coordinates/skycoord.rst]: Done

**Comment by maxnoe** [comment]
@eerovaher I think I addressed all comments and squashed into a single commit.

File: astropy/coordinates/baseframe.py
Comment (eerovaher) [inline astropy/coordinates/baseframe.py:706]: The `message` overwrites the default error message, so currently the users are not notified of the alternative they should be using.
```suggestion
        message="The {func}() {obj_type} is deprecated and may be removed in a future version. Use {alternative}() to obtain a dict of frame attribute names and default values. The fastest way to obtain the names is frame_attributes.keys()"
```

File: docs/changes/coordinates/13484.bugfix.rst
Comment (eerovaher) [inline docs/changes/coordinates/13484.bugfix.rst:5]: I think the current entry is too focused on implementation details. End users might not be too interested about some `dict` being slow to build.
```suggestion
``BaseCoordinateFrame.get_frame_attr_names()`` had a misleading name,
because it actually provided a ``dict`` of attribute names and
their default values. It is now deprecated and replaced by ``BaseCoordinateFrame.get_frame_attr_defaults()``.
The fastest way to obtain the attribute names is ``BaseFrame.frame_attributes.keys()``.
```

File: docs/coordinates/skycoord.rst
Comment (eerovaher) [inline docs/coordinates/skycoord.rst:550]: I'm not sure if this example should use `get_frame_attr_defaults()` or `frame_attributes`. I'm willing to merge either way.

**Comment by eerovaher** [comment]
pre-commit.ci autofix

**Review by eerovaher — CHANGES_REQUESTED** [review]
I think this is the last round of remarks I'll have. I would also appreciate it if the pull request would consist of a single commit which would also include the changes made by the pre-commit.ci bot.

File: docs/coordinates/skycoord.rst
Comment (maxnoe) [reply docs/coordinates/skycoord.rst:550]: Frame attr defaults is shown in the frame docs, so I'd show frame_attributes here

**Review by eerovaher — APPROVED** [review]
Thanks @maxnoe for your willingness to keep iterating this close to the deadline.

**Comment by maxnoe** [comment]
The codecov change is rather disturbing... (-10%), is that unrelated???

**Comment by maxnoe** [comment]
Ah nevermind, it seems codecov is uploaded from a partial test first and then when the full test suite finishes it's overwritten...

**Comment by maxnoe** [comment]
@eerovaher should this have been merged automatically? I saw you added the Label to merge after ci passes

**Comment by pllim** [comment]
> should this have been merged automatically?

This does not trigger any automation. It merely tells another core maintainer that this is ok to merge when CI passes, as sometimes a PR sits overnight but we have maintainers from a Round Earth™️ .

File: astropy/coordinates/baseframe.py
Comment (pllim) [inline astropy/coordinates/baseframe.py:707]: Just curious: Does f-string actually work on a deprecation warning message?

Comment (eerovaher) [reply astropy/coordinates/baseframe.py:707]: `message` is not an f-string. Field replacement is done by the `str.format()` method: https://github.com/astropy/astropy/blob/75c5e83d6344ca5aacb66794e79f3c55091c91ca/astropy/utils/decorators.py#L188-L193
See also the docstring: https://github.com/astropy/astropy/blob/75c5e83d6344ca5aacb66794e79f3c55091c91ca/astropy/utils/decorators.py#L36-L42

**Pull request merged**
